Fold activation permutation inside quantized conv operator

dzhulgakov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26339 Serialization for per channel qtensor
* #26243 Fix _empty_per_channel_affine_quantized to be less hacky
* **#26242 Fold activation permutation inside quantized conv operator**
* #26241 Fold weight permutation inside quantized conv operator
* #26240 Implement more support for per-channel quantization

According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.

Differential Revision: [D17443218](https://our.internmc.facebook.com/intern/diff/D17443218)

Commit:
permute for activations

Commit:
Update on "Fold activation permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.


[ghstack-poisoned]

Commit:
Update on "Fold activation permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.


[ghstack-poisoned]

Commit:
Update on "Fold activation permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.


[ghstack-poisoned]

jamesr66a commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -74,6 +74,15 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+    // mind. Ideally, we'd be compatible with conv2d behavior and preserve the
+    // inputs layout as is (doing necessary upconversions).
+    //
+    // However, to be more robust, we'd just force output layout to always be
```
"we'd just..." doesn't make it clear whether this is something you're proposing as a follow-up, or if this is how it's currently implemented

jamesr66a reviewed (commented):
Have you tested performance?

VitalyFedyunin commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -168,14 +177,17 @@ class QConv2dInt8 final : public c10::OperatorKernel {
-        outShape, device(kCPU).dtype(kQUInt8), output_scale, output_zero_point);
+        outShape, device(kCPU).dtype(kQUInt8), output_scale, output_zero_point,
+        MemoryFormat::ChannelsLast);
     auto buffer = at::zeros_like(output, output.options().dtype(at::kInt));
```
FYI: buffer here will be non channels last. Is it intended?

Commit:
Update on "Fold activation permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.

Differential Revision: [D17443218](https://our.internmc.facebook.com/intern/diff/D17443218)

[ghstack-poisoned]

dzhulgakov commented:
@jamesr66a - thanks for the benchmarking comment, I used benchmarks/operator_benchmarks and now they are the same before and after this diff. Also add a benchmark that has chained convolutions to make sure format is propagated through.

@VitalyFedyunin - there's currently a regression in `.contiguous(torch.channels_last)` call compared with permute. I kept it on permute (but inside C++), we can fix it later:

```
In [8]: orig = torch.empty((10,256,64,64))

In [9]: %timeit orig.contiguous(memory_format=torch.channels_last)
100 loops, best of 3: 8.61 ms per loop

In [10]: %timeit orig.permute([0,2,3,1]).contiguous()
100 loops, best of 3: 4.93 ms per loop
```

supriyar commented:
@dzhulgakov I submitted my PR for quantize conv, so you might want to rebase..

Commit:
Update on "Fold activation permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.

Differential Revision: [D17443218](https://our.internmc.facebook.com/intern/diff/D17443218)

[ghstack-poisoned]

Commit:
Update on "Fold activation permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.

Differential Revision: [D17443218](https://our.internmc.facebook.com/intern/diff/D17443218)

[ghstack-poisoned]

Commit:
Update on "Fold activation permutation inside quantized conv operator"


According to https://github.com/pytorch/pytorch/issues/19092 we always keep NCHW order and do handling inside the kernels. This PR fixes it for activations of the qconv by using MemoryLayout mechanism - activations stay logically as NCHW but strided as NHWC.

Note, that this version is more aggressive than eventual MemoryLayout mechanism - the QConv's output is always NHWC regardless of the input striding. I think it's ok as we don't have NCHW quantized kernels anyway - so the very first conv would magically switch the order, but I'm open to suggestions. Btw, it doesn't change behavior - same happens today in master because of the explicit permute() call.

Differential Revision: [D17443218](https://our.internmc.facebook.com/intern/diff/D17443218)

[ghstack-poisoned]
